[cpp, run] Using measurement of a qvector slice in return statement fails to compile

### Required prerequisites

- [x] Consult the [security policy](https://github.com/NVIDIA/cuda-quantum/security/policy). If reporting a security vulnerability, do not report the bug using this form. Use the process described in the policy to report the issue.
- [x] Make sure you've read the [documentation](https://nvidia.github.io/cuda-quantum/latest). Your issue may be addressed there.
- [x] Search the [issue tracker](https://github.com/NVIDIA/cuda-quantum/issues) to verify that this hasn't already been reported. +1 or comment there if it has.
- [ ] If possible, make a PR with a failing test to give us a starting point to work on!

### Describe the bug

The following return statement causes a compilation error:

```c++
__qpu__ bool test_bad() {
  cudaq::qvector qubits(6);
  x(qubits[3]);
  return mz(qubits.slice(3, 1))[0];
}
```

### Steps to reproduce the bug

```c++
#include <cudaq.h>
#include <iostream>

__qpu__ bool test_good() {
  cudaq::qvector qubits(6);
  x(qubits[3]);
  auto slice = qubits.slice(3, 1);
  return mz(slice)[0];
}

__qpu__ bool test_bad() {
  cudaq::qvector qubits(6);
  x(qubits[3]);
  return mz(qubits.slice(3, 1))[0];
}

int main() {
  {
    // Compiles
    auto results = cudaq::run(2, test_good);
    std::cout << "[";
    for (auto e: results) {
      std::cout << e << ", ";
    }
    std::cout << "]" << std::endl;
    // Expected: [1, 1, ]
  }
  {
    // Does not compile
    auto results = cudaq::run(2, test_bad);
    std::cout << "[";
    for (auto e: results) {
      std::cout << e << ", ";
    }
    std::cout << "]" << std::endl;
    // Expected: [1, 1, ]
  }
  return 0;
}
```

Output:
```
slice_bug.cpp:24:10: error: no matching function for call to 'mz'
  return mz(qubits.slice(3, 1))[0];
```

### Expected behavior

The `test_bad` code compiles and gives the same results as the `test_good` code.

### Is this a regression? If it is, put the last known working version (or commit) here.

Not a regression

### Environment

- **CUDA-Q version**: 
- **Python version**: 
- **C++ compiler**: 
- **Operating system**: 


### Suggestions

_No response_

--- a/runtime/cudaq/qis/qubit_qis.h
+++ b/runtime/cudaq/qis/qubit_qis.h
@@ -799,6 +799,15 @@ std::vector<measure_result> mz(QubitRange &q) {
   return b;
 }
 
+template <std::size_t Levels>
+std::vector<measure_result> mz(const qview<Levels> &q) {
+  std::vector<measure_result> b;
+  for (auto &qq : q) {
+    b.emplace_back(mz(qq));
+  }
+  return b;
+}
+
 template <typename... Qs>
 std::vector<measure_result> mz(qubit &q, Qs &&...qs);
 

--- a/runtime/cudaq/qis/qview.h
+++ b/runtime/cudaq/qis/qview.h
@@ -41,9 +41,11 @@ public:
 
   /// @brief Iterator interface, begin.
   auto begin() { return qudits.begin(); }
+  auto begin() const { return qudits.begin(); }
 
   /// @brief Iterator interface, end.
   auto end() { return qudits.end(); }
+  auto end() const { return qudits.end(); }
 
   /// @brief Returns the qudit at `idx`.
   value_type &operator[](const std::size_t idx) { return qudits[idx]; }

--- a/test/AST-Quake/slice.cpp
+++ b/test/AST-Quake/slice.cpp
@@ -30,3 +30,9 @@ struct SliceTest {
 // CHECK:           call @{{.*}}other{{.*}}(%[[VAL_14]]) : (!quake.veq<?>) -> ()
 // CHECK:           return
 // CHECK:         }
+
+__qpu__ bool issue_3092() {
+  cudaq::qvector qubits(6);
+  x(qubits[3]);
+  return mz(qubits.slice(3, 1))[0];
+}

--- a/test/AST-Quake/slice.cpp
+++ b/test/AST-Quake/slice.cpp
@@ -6,7 +6,7 @@
  * the terms of the Apache License 2.0 which accompanies this distribution.    *
  ******************************************************************************/
 
-// RUN: cudaq-quake %cpp_std %s | FileCheck %s
+// RUN: cudaq-quake %cpp_std %s | cudaq-opt | FileCheck %s
 
 #include <cudaq.h>
 
@@ -36,3 +36,16 @@ __qpu__ bool issue_3092() {
   x(qubits[3]);
   return mz(qubits.slice(3, 1))[0];
 }
+
+// CHECK-LABEL:   func.func @__nvqpp__mlirgen__function_issue_3092._Z10issue_3092v() -> i1
+// CHECK:           %[[VAL_0:.*]] = quake.alloca !quake.veq<6>
+// CHECK:           %[[VAL_1:.*]] = quake.extract_ref %[[VAL_0]][3] : (!quake.veq<6>) -> !quake.ref
+// CHECK:           quake.x %[[VAL_1]] : (!quake.ref) -> ()
+// CHECK:           %[[VAL_2:.*]] = quake.subveq %[[VAL_0]], 3, 3 : (!quake.veq<6>) -> !quake.veq<1>
+// CHECK:           %[[VAL_3:.*]] = quake.mz %[[VAL_2]] : (!quake.veq<1>) -> !cc.stdvec<!quake.measure>
+// CHECK:           %[[VAL_4:.*]] = quake.discriminate %[[VAL_3]] : (!cc.stdvec<!quake.measure>) -> !cc.stdvec<i1>
+// CHECK:           %[[VAL_5:.*]] = cc.stdvec_data %[[VAL_4]] : (!cc.stdvec<i1>) -> !cc.ptr<!cc.array<i1 x ?>>
+// CHECK:           %[[VAL_6:.*]] = cc.cast %[[VAL_5]] : (!cc.ptr<!cc.array<i1 x ?>>) -> !cc.ptr<i1>
+// CHECK:           %[[VAL_7:.*]] = cc.load %[[VAL_6]] : !cc.ptr<i1>
+// CHECK:           return %[[VAL_7]] : i1
+// CHECK:         }

